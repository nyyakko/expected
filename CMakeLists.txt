cmake_minimum_required(VERSION 3.25)

project(expected LANGUAGES CXX)

set(EXPORT_DIR "${CMAKE_SOURCE_DIR}/build/cmake")

include(cmake/CPM.cmake)
include(cmake/static_analyzers.cmake)
include(GNUInstallDirs)

# set(tl-expected_CompilerOptions ${tl-expected_CompilerOptions})
# set(tl-expected_LinkerOptions ${tl-expected_LinkerOptions})

# set(tl-expected_ExternalLibraries)

add_subdirectory(expected)
