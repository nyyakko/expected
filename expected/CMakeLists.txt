add_subdirectory(include/tl)

add_library(${PROJECT_NAME} INTERFACE ${expected_HeaderFiles})
add_library(tl::expected ALIAS ${PROJECT_NAME})

if (ENABLE_CLANGTIDY)
    enable_clang_tidy(${PROJECT_NAME})
endif()

if (ENABLE_CPPCHECK)
    enable_cppcheck(${PROJECT_NAME})
endif()

target_include_directories(${PROJECT_NAME}
    INTERFACE
        "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
        "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>"
)

set_target_properties(${PROJECT_NAME} PROPERTIES VERIFY_INTERFACE_HEADER_SETS TRUE)

target_compile_features(${PROJECT_NAME} INTERFACE cxx_std_23)

target_sources(${PROJECT_NAME} PUBLIC FILE_SET HEADERS BASE_DIRS include FILES ${expected_HeaderFiles})

export(TARGETS      ${PROJECT_NAME}
       FILE         ${EXPORT_DIR}/${PROJECT_NAME}Targets.cmake
       NAMESPACE    tl::
)

install(TARGETS     ${PROJECT_NAME}
        EXPORT      ${PROJECT_NAME}Targets
        ARCHIVE
        FILE_SET HEADERS
        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
)

install(EXPORT      ${PROJECT_NAME}Targets
        DESTINATION ${CMAKE_INSTALL_LIBDIR}/${PROJECT_NAME}/cmake
        NAMESPACE   tl::
)

install(FILES ${PROJECT_SOURCE_DIR}/cmake/${PROJECT_NAME}Config.cmake
        DESTINATION ${CMAKE_INSTALL_LIBDIR}/${PROJECT_NAME}/cmake
)

target_compile_options(${PROJECT_NAME} INTERFACE ${expected_CompilerOptions})
target_link_libraries(${PROJECT_NAME} INTERFACE ${expected_ExternalLibraries})
